packages/api/lib/ems.js

```diff
@@ -0,0 +1,280 @@
+            {
+              range: {
+                [`${timeFieldName}`]: {
+                  gte: moment.utc(startTime).toDate().getTime(),
+                  lt: moment.utc(endTime).toDate().getTime()
+                }
+              }
+            },
+            {
+              terms: {
+                // filter out 'running' status
+                status: ['failed', 'completed']
+              }
+            }]
+        }
+      }
+    }
+  };
+  if (type === 'deletedgranule') params._source = ['granuleId', 'deletedAt'];
+  return params;
+}
+
+/**
+ * get the value of EMS record field from corresponding field of the  granule record
+ *
+ * @param {Object} granule - es granule record
+ * @param {string} emsField - EMS field
+ * @param {string} granField - granule field
+ * @returns {string} granule metadata for EMS
+ */
+function getEmsFieldFromGranField(granule, emsField, granField) {
+  const metadata = get(granule, granField, null);
+  let result = metadata;
+  switch (emsField) {
+  case 'product':
+    result = deconstructCollectionId(metadata).name;
+    break;
+  case 'versionID':
+    result = parseInt(deconstructCollectionId(metadata).version, 10);
+    break;
+  case 'deleteFromArchive':
+    result = 'N';
+    break;
+  case 'totalFiles':
+    result = metadata.length;
+    break;
+  case 'productState':
+    result = (metadata === 'completed') ? 'Successful' : 'Failed';
+    break;
+  // datetime format YYYY-MM-DD HH:MMAMorPM GMT
+  case 'deleteEffectiveDate':
+  case 'insertTime':
+    // milliseconds to string
+    result = (metadata) ? moment.utc(new Date(metadata)).format('YYYY-MM-DD hh:mmA') : metadata;
+    break;
+  case 'lastUpdate':
+  case 'processingStartDateTime':
+  case 'processingEndDateTime':
+  case 'beginningDateTime':
+  case 'endingDateTime':
+  case 'productionDateTime':
+    // string to different format string
+    result = (metadata) ? moment.utc(Date.parse(metadata)).format('YYYY-MM-DD hh:mmA') : metadata;
+    break;
+  default:
+    break;
+  }
+  return result;
+}
+
+/**
+ * build EMS records from es granules
+ *
+ * @param {Object} mapping - mapping of EMS fields to granule fields
+ * @param {Object} granules - es granules
+ * @returns {Array<string>} EMS records
+ */
+function buildEMSRecords(mapping, granules) {
+  const records = granules.map((g) => {
+    const record = Object.keys(mapping).map((k) => getEmsFieldFromGranField(g, k, mapping[k]));
+    return record.join('|&|');
+  });
+  return records;
+}
+
+/**
+ * upload a report to s3, a rev file is created if the report already exists in s3
+ *
+ * @param {string} filename - file to be upload to s3
+ * @returns {string} - uploaded file in s3
+ */
+async function uploadReportToS3(filename) {
+  const bucket = process.env.bucket;
+  const originalKey = `${process.env.stackName}/ems/${path.basename(filename)}`;
+  let key = originalKey;
+  let exists = await aws.fileExists(bucket, key);
+  let i = 1;
+  while (exists) {
```

Comment:
Instead of repeatedly calling `fileExists`, you could make a single call to `aws.s3().listObjectsV2()` with a prefix set to determine what key to use.

Also, this function would be easier to read if you extracted the "figure out what key to use" functionality into a separate function that this function called.